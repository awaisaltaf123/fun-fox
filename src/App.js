import React, { useState, useEffect } from "react";
import TaskModal from "./components/TaskModal";
import TaskList from "./components/TaskList";
import { DragDropContext } from "react-beautiful-dnd";
import { fetchUsersTasks, fetchUsersGroup } from "./api/taskManagement";
import { notifySuccess } from "./Helper/notifyMessages";
import { filterGroupAndExtractIds } from "./Helper/filterData";
import Spinner from "./Helper/Spinner/Spinner";

const App = () => {
  const [tasks, setTasks] = useState([]);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    Promise.all([fetchUsersGroup(), fetchUsersTasks()])
      .then(([groupData, taskData]) => {
        const groupId = process.env.REACT_APP_GROUP_ID;
        const filteredTasks = filterGroupAndExtractIds(
          groupData,
          taskData,
          groupId
        );
        setTasks(filteredTasks);
        setIsLoading(false);
      })
      .catch((error) => {
        setIsLoading(false);
        console.log("error when fetching Data!!", error);
      });
  }, []);

  const handleAddTask = (newTask) => {
    const newTaskWithId = {
      ...newTask,
      id: Date.now(),
      completed: false,
      dragableId: Math.random().toString(),
    };
    setTasks([...tasks, newTaskWithId]);
    notifySuccess("Item added successfully!");
  };

  const handleDeleteTask = (taskId) => {
    setTasks(tasks.filter((task) => task.id !== taskId));
    notifySuccess("Item Deleted!");
  };

  const handleCompleteTask = (taskId) => {
    setTasks(
      tasks.map((task) =>
        task.id === taskId ? { ...task, completed: !task.completed } : task
      )
    );
    notifySuccess("Item status Changed!!");
  };

  const handleOnDragEnd = (result) => {
    if (!result.destination) return;

    const items = Array.from(tasks);
    const [reorderedItem] = items.splice(result.source.index, 1);
    items.splice(result.destination.index, 0, reorderedItem);

    setTasks(items);
  };

  return (
    <div>
      {isLoading && <Spinner />}
      <h1 className="text-xl font-bold pb-10 pt-5 ml-2 flex items-center justify-center">
        Task Management System
      </h1>
      <TaskModal onAddTask={handleAddTask} />
      <DragDropContext onDragEnd={handleOnDragEnd}>
        <TaskList
          tasks={tasks}
          onDelete={handleDeleteTask}
          onComplete={handleCompleteTask}
        />
      </DragDropContext>
    </div>
  );
};

export default App;
