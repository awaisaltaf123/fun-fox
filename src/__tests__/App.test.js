import React from "react";
import { render } from "@testing-library/react";
import App from "../App";

describe("App", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render correctly", () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });
});
