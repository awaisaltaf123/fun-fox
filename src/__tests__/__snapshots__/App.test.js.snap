// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
<div>
  <div>
    <div
      class="circular-spinner"
    >
      <div
        class="spinner-circle"
      />
    </div>
    <h1
      class="text-xl font-bold pb-10 pt-5 ml-2 flex items-center justify-center"
    >
      Task Management System
    </h1>
    <div
      class="flex items-end justify-end"
    >
      <button
        class="bg-blue-500 text-white px-4 py-2 rounded-md mr-10"
      >
        Add New Task
      </button>
    </div>
    <div
      data-rbd-droppable-context-id="0"
      data-rbd-droppable-id="droppable"
    />
  </div>
</div>
`;
